src/lib/settings/group/SettingsArray.js

```diff
@@ -0,0 +1,127 @@
+	}
+
+	includes(value) {
+		return this.data.includes(value);
+	}
+
+	async update(values, options) {
+		if (!Array.isArray(values)) values = [values];
+
+		const guild = resolveGuild(this.base.gateway.client, 'guild' in options ? options.guild : this.base.target);
+		const { entry } = this;
+		let indexing = false;
+
+		// Not sure of a better way to do this, come back at a later time
+		if (values.some(checkForIndex)) {
+			if (!(values.every(checkForIndex))) throw "Indexing found. You must only use straight indexing or no indexing, not a mixture of both.";
+			indexing = true;
+		}
+
+		const { errors, clone } = await this._parse(values, options, guild, indexing);
+		// This might need to be changed/adjusted to give the user better throw behavior.
+		if (errors.length) throw { errors, updated: [] };
+		if (arraysStrictEquals(this.data, clone)) return { errors: [], updated: [] };
+		
+		const result = [{ key: entry.path, value: clone, entry }];
+		this._save(result);
+		
+		return { errors, updated: result };
+	}
+
+	async _parse(values, options, guild, indexing) {
+		const { serializer: { serialize } } = this.entry;
+
+		if (indexing) {
+			values = await Promise.all(values.map(async ([i, v]) => ([i, serialize(await entry.parse(v, guild))])));
+		} else if (!Array.isArray(values)) {
+			values = serialize(await entry.parse(values, guild));
+		} else values = await Promise.all(values.map(async val => serialize(await entry.parse(val, guild))));
+
+		const { action = 'auto' } = options;
+		if (!indexing && action === 'overwrite') return values;
+
+		const clone = this.get();
+
+		// Errors are given back in the order that values were sent in
+		const errors = [];
+
+		// This value has an index paired with it
+		if (indexing) {
+			for (const val of values) {
+				let index = val[0];
+				if (clone.length === 0 && index > 0) {
+					errors.push({ input: val, message: 'The current array is empty. The index must start at 0.' });
+				} else if (index < 0 || index > clone.length + 1) {
+					errors.push({ input: val, message: `The index ${index} is bigger than the current array. It must be a value in the range of 0..${clone.length + 1}.` });
+				}
+				clone[index] = val[1];
+			}
+		} else if (action === 'auto') {
+			for (const val of values) {
+				const index = clone.indexOf(val);
+				if (index === -1) clone.push(val);
+				else clone.splice(index, 1)
+			}
+		} else if (action === 'add') {
+			for (const val of values) {
+				if (clone.includes(val)) errors.push({ input: val, message: `The value ${val} for the key ${entry.path} already exists.` });
+				else clone.push(val);
+			}
+		} else if (action === 'remove') {
+			for (const val of values) {
+				const index = clone.indexOf(val);
+				if (index === -1) errors.push({ input: val, message: `The value ${val} for the key ${entry.path} does not exist.` });
+				else clone.splice(index, 1);
+			}
+		} else {
+			throw `The ${action} array action is not a valid SettingsUpdateArrayAction.`;
```

Comment:
Also, this doesn't handle upper case... Should we handle that too? probably not...

```suggestion
			throw `The "${action}" array action is not a valid SettingsUpdateArrayAction.`;
```